File: C++_Version/206Reverse_Linked_List.cpp
//Reverse a singly linked list.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
using namespace std;


//* Definition for singly-linked list.
struct ListNode 
{
	int val;
	ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* reverseList(ListNode* head) {
		ListNode *pre = NULL, *now = head, *next;

		while (now != NULL)
		{
			next = now->next;
			now->next = pre;
			pre = now;
			now = next;
		}

		return pre;
	}
};

int main()
{
	ListNode* headA = new ListNode(0), *ptemp1, *ptemp2 = headA ;

	int a[] = { 0,1,2,6,3,4,5,6 };

	for (int i = 1; i < 8; ++i)
	{
		ptemp1 = new ListNode(a[i]);
		ptemp2->next = ptemp1;
		ptemp2 = ptemp1;
	}


	Solution sol1;
	headA = sol1.reverseList(headA);

	system("pause");

	return 0;
}

File: Java_Version/206_Reverse_Linked_List.java
/*
  Reverse a singly linked list.

Example:

        Input: 1->2->3->4->5->NULL
        Output: 5->4->3->2->1->NULL

Follow up:
  A linked list can be reversed either iteratively or recursively. Could you
implement both?
*/



class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}



class Solution {

    public static void main(String[] args) {

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);

        ListNode newHead = new Solution().reverseList(head);

        System.out.println();
    }

    public ListNode reverseList(ListNode head) {

        return reverseListNonRecursively(head);
        // return reverseListRecursively(head);
    }

    private ListNode reverseListNonRecursively(ListNode head) {

        if (head == null || head.next == null) {
            return head;
        }

        ListNode lastNode = null, currentNode = head;
        while (currentNode != null) {
            ListNode nextNode = currentNode.next;
            currentNode.next = lastNode;

            lastNode = currentNode;
            currentNode = nextNode;
        }

        return lastNode;
    }

    private ListNode reverseListRecursively(ListNode head) {

        if (head == null) {
            return null;
        } else if (head.next == null) {
            return head;
        }

        ListNode nextNode = head.next;
        head.next = null;
        ListNode newHead = reverseListRecursively(nextNode);
        nextNode.next = head;
        return newHead;
    }
}